docs: about time for plugins (#2169)
sopel-irc/sopel: docs/source/plugin/bot.rst

    Read about the :class:`~sopel.tools.target.Channel` and
    :class:`~sopel.tools.target.User` classes for more details.

About time
==========

Your plugin may want to display dates and times in messages. For that, you can
always count on the :mod:`datetime` built-in module. However, what if you would

> like to respect the date-format for a given user or a given channel?
> :mod:`Sopel<sopel.tools.time>` can help you with that with various functions:

Reviewer: ```suggestion
like to respect the date-format for a given user or a given channel? Various
functions of :mod:`Sopel<sopel.tools.time>` can help you with that:
```

* :func:`~sopel.tools.time.get_timezone` will fetch the right timezone for you
* :func:`~sopel.tools.time.format_time` will format your aware datetime for you

Here is a full example of that::

    import datetime

>     import pytz

Reviewer: ```suggestion
    import pytz  # always available; Sopel requires it
```

LMK if you think it's _more_ confusing to call attention to this.

Author: I'd rather put a note at the end than in the code sample.

Reviewer: That's fair.

    from sopel import plugin

>     from sopel.tools.time import format_time, get_timezone

Reviewer: Is this a good time (ha!) to demonstrate our suggested import style, `from sopel.tools import time`? Or maybe not, since there's a stdlib module named `time` too. 🙁

Author: That, and also I'd like to keep things short for the example.

    @plugin.command('.t')
    @plugin.require_chanmsg
    def my_command(bot, trigger):
        """Give time in a channel."""
        time = pytz.UTC.localize(datetime.datetime.utcnow())
        timezone = get_timezone(
            bot.db,
            bot.settings,
            nick=trigger.nick,
            channel=trigger.sender,
        )

>         formated_time = format_time(
>             bot.db,
>             bot.settings,
>             timezone,
>             trigger.nick,
>             trigger.sender,
>             time,
>         )
>         bot.say(formated_time)

Reviewer: ```suggestion
        formatted_time = format_time(
            bot.db,
            bot.settings,
            timezone,
            trigger.nick,
            trigger.sender,
            time,
        )
        bot.say(formatted_time)
```

Getting the time
----------------

> As mentionned earlier, Sopel relies on a Python built-in module:

Reviewer: ```suggestion
As mentioned earlier, Sopel relies on a Python built-in module:
```

:mod:`datetime`. This module allows you to get the current time like this::

    >>> import datetime
    >>> datetime.datetime.now()
    datetime.datetime(2021, 7, 26, 18, 7, 13, 491786)
    >>> datetime.datetime.utcnow()
    datetime.datetime(2021, 7, 26, 16, 7, 16, 496404)

As you can see at the moment of writing this documentation, there was a 2h

> offset between the local time and UTC. To properly manage timezone and UTC
> offset, it is best to rely on ``pytz``, and work with UTC only::

Reviewer: ```suggestion
offset between the local time and UTC. To properly manage timezones and UTC
offsets, it is best to rely on ``pytz``, and work with UTC only::
```

    >>> import pytz
    >>> pytz.UTC.localize(datetime.datetime.utcnow())
    datetime.datetime(2021, 7, 26, 16, 7, 19, 321828, tzinfo=<UTC>)

> This way, you'll always have an aware datetime that doesn't depend on your
> local time to work with (store, compare, manipulate, etc.) and when you need it
> you will convert it to the proper timezone.

Reviewer: ```suggestion
This way, you'll always have an aware datetime to work with (store, compare,
manipulate, etc.) that doesn't depend on your local time, and you'll only
convert it to the proper timezone when you need to.
```

.. note::

>     You should **always** work with aware datetime, and it's easier to always
>     work with UTC+0 datetime as input and storage then to convert into another
>     timezone when displaying time to a user.

Reviewer: ```suggestion
    You should **always** work with aware datetime objects. It's also easier to
    always work with UTC+0 datetime as input and in storage, then convert to
    another timezone when displaying time to a user.
```

    >>> tz = pytz.timezone('Europe/Paris')
    >>> now = pytz.UTC.localize(datetime.datetime.utcnow())
    >>> now.astimezone(tz)
    datetime.datetime(
        2021, 7, 26, 18, 7, 49, 27970,
        tzinfo=<DstTzInfo 'Europe/Paris' CEST+2:00:00 DST>)
    >>> tz = pytz.timezone('America/Chicago')
    >>> now.astimezone(tz)
    datetime.datetime(
        2021, 7, 26, 11, 7, 58, 610998,
        tzinfo=<DstTzInfo 'America/Chicago' CDT-1 day, 19:00:00 DST>

.. __: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

To get the IANA timezone for a given user or channel, you should use the
:func:`~sopel.tools.time.get_timezone` function::

    >>> from sopel.tools.time import get_timezone
    >>> # assuming bot is an instance of sopel.bot.Sopel
    >>> custom_tz = get_timezone(
    ...     bot.db, bot.settings,
    ...     zone=None, nick='Nick', channel='#sopel',
    ... )  # should be something like "Europe/Paris"
    >>> local_now = now.astimezone(pytz.timezone(custom_tz))

This function does all the heavy lifting of looking for the right timezone, as
configured for a user, a channel, or the bot itself.

Format time
-----------

So far, you have:

> * an UTC+0 aware datetime

Reviewer: ```suggestion
* an aware datetime in UTC+0
```

Or:

```suggestion
* an aware UTC+0 datetime
```

:mod:`datetime`. This module allows you to get the current time like this::

    >>> import datetime
    >>> datetime.datetime.now()
    datetime.datetime(2021, 7, 26, 18, 7, 13, 491786)
    >>> datetime.datetime.utcnow()
    datetime.datetime(2021, 7, 26, 16, 7, 16, 496404)

As you can see at the moment of writing this documentation, there was a 2h
offset between the local time and UTC. To properly manage timezones and UTC
offsets, it is best to rely on ``pytz``, and work with UTC only::

    >>> import pytz
    >>> pytz.UTC.localize(datetime.datetime.utcnow())
    datetime.datetime(2021, 7, 26, 16, 7, 19, 321828, tzinfo=<UTC>)

This way, you'll always have an aware datetime to work with (store, compare,
manipulate, etc.) that doesn't depend on your local time, and you'll only
convert it to the proper timezone when you need to.

.. note::

    You should **always** work with aware datetime objects. It's also easier to
    always work with UTC+0 datetime as input and in storage, then convert to
    another timezone when displaying time to a user.

Getting the timezone
--------------------

Sopel uses ``pytz`` to handle timezones to manipulate aware datetimes, i.e.

> datetimes with a timezone information such as the UTC Offset and DST
> (`Daylight Saving Time`__). When using this library, getting a timezone

Reviewer: ```suggestion
datetimes with timezone-related information such as the UTC Offset and DST
(`Daylight Saving Time`__) status. When using this library, getting a timezone
```

Not strictly necessary to go full "timezone-related", but the "a" is not valid syntax.

Might be smoother to link "DST" directly to the wiki page instead of stopping to define it in a parenthetical. (Could use [the `:abbr:` role](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-abbr), though it would have to be tested; Furo doesn't appear to include that in [its Kitchen Sink](https://pradyunsg.me/furo/kitchen-sink/).)

is straightforward (as long as you know the `IANA name of said timezone`__)::

    >>> tz = pytz.timezone('Europe/Paris')

Then you can get an UTC datetime::

    >>> now = pytz.UTC.localize(datetime.datetime.utcnow())

Which you can convert to your timezone, or to another timezone::

    >>> now.astimezone(tz)  # convert to timezone
    datetime.datetime(
        2021, 7, 26, 18, 7, 49, 27970,
        tzinfo=<DstTzInfo 'Europe/Paris' CEST+2:00:00 DST>)
    >>> tz = pytz.timezone('America/Chicago')

>     >>> now.astimezone(tz)  # convert to a different timezone

Reviewer: Would it be clearer to use a different variable name instead of overriding the previously defined `tz`?

Author: Not sure. We should ask @xnaas if it's confusing or not. :wink:

Reviewer: In the absence of feedback from @xnaas, let's say the first variable should be `paris` and this second one named `chicago`, because my gut says someone is going to read quickly and get confused if they have the same name.

As you can see at the moment of writing this documentation, there was a 2h
offset between the local time and UTC. To properly manage timezones and UTC
offsets, it is best to rely on ``pytz``, and work with UTC only::

    >>> import pytz
    >>> pytz.UTC.localize(datetime.datetime.utcnow())
    datetime.datetime(2021, 7, 26, 16, 7, 19, 321828, tzinfo=<UTC>)

This way, you'll always have an aware datetime to work with (store, compare,
manipulate, etc.) that doesn't depend on your local time, and you'll only
convert it to the proper timezone when you need to.

.. note::

    You should **always** work with aware datetime objects. It's also easier to
    always work with UTC+0 datetime as input and in storage, then convert to
    another timezone when displaying time to a user.

Getting the timezone
--------------------

Sopel uses ``pytz`` to handle timezones to manipulate aware datetimes, i.e.
datetimes with a timezone information such as the UTC Offset and DST
(`Daylight Saving Time`__). When using this library, getting a timezone
is straightforward (as long as you know the `IANA name of said timezone`__)::

    >>> tz = pytz.timezone('Europe/Paris')

> Then you can get an UTC datetime::

Reviewer: ```suggestion
Then you can get a UTC datetime::
```

"an" if UTC was read as "uhtk", sure. But most people are going to read it as an initialism and say "you tee see", so it'll be "a" in practice.

Don't you love English? 🙃

Author: No, I don't.
